app/util/appUtil.js

```diff
@@ -75,21 +91,77 @@ function formatHash(filestoreResponse) {
-  return Object.fromEntries(
+  const metadataItems = Object.entries(metadata)
+  if (metadataItems.length > MAX_METADATA_COUNT)
+    throw new Error(`Metadata has too many items: ${metadataItems.length}. Max item count: ${MAX_METADATA_COUNT}`)
+
+  return new Map(
     await Promise.all(
-      Object.entries(metadata).map(async ([key, filename]) => {
-        if (key.length > METADATA_KEY_LENGTH)
-          throw new Error(`Key: ${key} is too long. Maximum key length is ${METADATA_KEY_LENGTH}`)
-
-        const file = files[filename]
-        if (!file) throw new Error(`Error no attached file found for ${filename}`)
-        const filestoreResponse = await addFile(file)
-        return [key, formatHash(filestoreResponse)]
+      metadataItems.map(async ([key, value]) => {
+        const lengthInHex = new TextEncoder('hex').encode(key).length
+        if (lengthInHex > METADATA_KEY_LENGTH)
+          throw new Error(`Key: ${key} is too long: ${lengthInHex} bytes (hex). Max length: ${METADATA_KEY_LENGTH}`)
+
+        const validMetadataValueTypes = Object.keys(apiOptions.types.MetadataValue._enum)
+        if (typeof value !== 'object' || !validMetadataValueTypes.some((type) => type.toUpperCase() === value.type)) {
+          throw new Error(
+            `Error invalid type in ${key}:${JSON.stringify(value)}. Must be one of ${validMetadataValueTypes.map((t) =>
+              t.toUpperCase()
+            )}`
+          )
+        }
+
+        switch (value.type) {
+          case 'LITERAL':
+            value = processLiteral(value)
+            break
+          case 'FILE':
+            value = await processFile(value, files)
+            break
+          default:
+          case 'NONE':
+            value = { None: null }
+            break
+        }
+
+        const keyAsUint8Array = utf8ToUint8Array(key, METADATA_KEY_LENGTH)
+        return [keyAsUint8Array, value]
       })
     )
   )
 }
 
+const processLiteral = (value) => {
+  const literalValue = value.value
+  if (!literalValue) throw new Error(`Literal metadata requires a value field`)
+
+  const lengthInHex = new TextEncoder('hex').encode(literalValue).length
+  if (lengthInHex > METADATA_VALUE_LITERAL_LENGTH)
+    throw new Error(
+      `${literalValue} is too long: ${lengthInHex} bytes (hex). Maximum length is ${METADATA_VALUE_LITERAL_LENGTH}`
+    )
+
+  const valueAsUint8Array = utf8ToUint8Array(literalValue, METADATA_VALUE_LITERAL_LENGTH)
+  return { Literal: valueAsUint8Array }
+}
+
+const processFile = async (value, files) => {
+  if (!value.value) throw new Error(`File metadata requires a value field`)
+
+  const filePath = value.value
+  const file = files[filePath]
+  if (!file) throw new Error(`Error no attached file found for ${filePath}`)
+
+  const filestoreResponse = await addFile(file)
+  return { File: formatHash(filestoreResponse) }
+}
+
+const utf8ToUint8Array = (str, len) => {
+  const arr = new Uint8Array(len)
+  arr.set(Buffer.from(str))
```

Comment:
I think `Buffer.from` here should probably be passed a `utf8` param to make it clear how the string will be interpretted

Reply:
added `utf8`